File: src/main/java/com/sfh/pokeRogueBot/service/ShortTermMemory.java
package com.sfh.pokeRogueBot.service;

import com.sfh.pokeRogueBot.model.exception.ActionLoopDetectedException;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;

@Component
public class ShortTermMemory {

    private final int memorySize;
    private final int minPhasesForLoop;

    List<String> lastMemories = new LinkedList<>();

    public ShortTermMemory(
            @Value("${brain.shortTermMemory.memorySize}") int memorySize,
            @Value("${brain.shortTermMemory.minPhasesForLoop}") int minPhasesForLoop
    ) {
        this.memorySize = memorySize;
        this.minPhasesForLoop = minPhasesForLoop;
    }

    public void memorizePhase(String memory) throws ActionLoopDetectedException {
        if(lastMemories.size() < memorySize) {
            lastMemories.add(memory);
        } else {
            checkForActionLoop();
            lastMemories.remove(0);
            lastMemories.add(memory);
        }
    }

    public void clearMemory() {
        lastMemories.clear();
    }

    private void checkForActionLoop() throws ActionLoopDetectedException {
        Map<String, Integer> phaseCount = new java.util.HashMap<>();
        for(int i = 0; i < memorySize; i++) {
            if(phaseCount.containsKey(lastMemories.get(i))) {
                phaseCount.put(lastMemories.get(i), phaseCount.get(lastMemories.get(i)) + 1);
            } else {
                phaseCount.put(lastMemories.get(i), 1);
            }
        }

        if(phaseCount.size() < minPhasesForLoop) {
            StringJoiner sj = new StringJoiner(", ");
            phaseCount.forEach((k, v) -> sj.add(k + ": " + v));
            throw new ActionLoopDetectedException("Action loop detected: " + sj);
        }
    }
}

File: src/main/kotlin/com/sfh/pokeRogueBot/service/ShortTermMemory.kt
package com.sfh.pokeRogueBot.service

import com.sfh.pokeRogueBot.model.exception.ActionLoopDetectedException
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.*

/**
 * ShortTermMemory is used to store the last actions of the bot in the current application run.
 * It helps to detect action loops by checking if the same actions are repeated too often.
 */
@Component
class ShortTermMemory(
    @Value("\${brain.shortTermMemory.memorySize}") private val lastPhasesMemorySize: Int,
    @Value("\${brain.shortTermMemory.minPhasesForLoop}") private val minPhasesForLoop: Int,
) {

    private val lastPhasesMemories: MutableList<String> = LinkedList()

    fun memorizePhase(memory: String) {
        if (lastPhasesMemories.size < lastPhasesMemorySize) {
            lastPhasesMemories.add(memory)
        } else {
            checkForActionLoop()
            lastPhasesMemories.removeFirst()
            lastPhasesMemories.add(memory)
        }
    }

    fun clearLastPhaseMemory() {
        lastPhasesMemories.clear()
    }

    private fun checkForActionLoop() {
        val phaseCount = mutableMapOf<String, Int>()

        for (i in 0 until lastPhasesMemorySize) {
            val phase = lastPhasesMemories[i]
            phaseCount[phase] = phaseCount.getOrDefault(phase, 0) + 1
        }

        if (phaseCount.size < minPhasesForLoop) {
            val phaseCountString = phaseCount.entries.joinToString(", ") { "${it.key}: ${it.value}" }
            throw ActionLoopDetectedException("Action loop detected: $phaseCountString")
        }
    }
}